### PR TITLE
printf: Move validation inside instrumentation time

### PR DESCRIPTION
**Pull request by spencer-lunarg**
This 

1. Adds a way to do `LogWarning`/`LogExternal` inside shader instrumentation
2. Moves the DebugPrintf validation (https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8294/files) from core checks to inside when we do the actually shader instrumentation of debug printf

### CODE REVIEW

File: layers/gpu/spirv/debug_printf_pass.cpp [hunk 1]
```diff
@@ -497,5 +499,264 @@ bool DebugPrintfPass::Run() {
 
 void DebugPrintfPass::PrintDebugInfo() { std::cout << "DebugPrintfPass\n\tinstrumentation count: " << instrumented_count_ << '\n'; }
 
+// Strictly speaking - the format given in GLSL_EXT_debug_printf is a client side implementation of SPIR-V
+// NonSemantic.DebugPrintf There is nothing someone from creating a debug printf implementation that goes `printf("Use this &q
+// to print int", myInt)` But this requires both having a different HLL and Tool consuming it. Currently RenderDoc and the
+// Validation Layers both follow the same syntax, but that also could possibly change.
+//
+// Therefore, we validate these here based on the VVL implementation only
+bool DebugPrintfPass::Validate(const Function& current_function) {
```
Comment (spencer-lunarg) [inline layers/gpu/spirv/debug_printf_pass.cpp:508]: this was just copy-and-pasting and having to adjust the `LogWarning` to use `InternalWarning`

File: layers/gpu/spirv/module.cpp [hunk 2]
```diff
@@ -637,8 +639,21 @@ void Module::PostProcess() {
     }
 }
 
-// TODO - Currently this is not tied into validation layers callback, should be using LogWarning/LogError
-void InternalWarning(const char* message) { std::cout << "[Internal Shader Instrumentation Warning] " << message << '\n'; }
+void Module::InternalWarning(const char* tag, const char* message) {
+    if (debug_report_) {
+        debug_report_->DebugLogMsg(kWarningBit, {}, message, tag);
```
Comment (spencer-lunarg) [inline layers/gpu/spirv/module.cpp:644]: using `DebugLogMsg` for now, we need some re-working how to make the variadic function more easily portable without having to bring in the entire `chassis.h` file

File: layers/gpu/spirv/debug_printf_pass.cpp [hunk 3]
```diff
@@ -497,5 +499,264 @@ bool DebugPrintfPass::Run() {
 
 void DebugPrintfPass::PrintDebugInfo() { std::cout << "DebugPrintfPass\n\tinstrumentation count: " << instrumented_count_ << '\n'; }
 
+// Strictly speaking - the format given in GLSL_EXT_debug_printf is a client side implementation of SPIR-V
+// NonSemantic.DebugPrintf There is nothing someone from creating a debug printf implementation that goes `printf("Use this &q
```
Comment (arno-lunarg) [inline layers/gpu/spirv/debug_printf_pass.cpp:503]:
```suggestion
// NonSemantic.DebugPrintf There is nothing stopping someone from creating a debug printf implementation that goes `printf("Use this &q
```

File: layers/gpu/spirv/module.cpp [hunk 4]
```diff
@@ -637,8 +639,21 @@ void Module::PostProcess() {
...
+    } else {
+        std::cout << "[" << tag << "]" << message << '\n';
+    }
+}
+
+void Module::InternalError(const char* tag, const char* message) {
+    if (debug_report_) {
+        debug_report_->DebugLogMsg(kErrorBit, {}, message, tag);
+    } else {
+        std::cerr << "[" << tag << "]" << message << '\n';
```
Comment (arno-lunarg) [inline layers/gpu/spirv/module.cpp:654]:
```suggestion
        std::cerr << "[" << tag << "] " << message << '\n';
```

File: layers/error_message/logging.cpp [hunk 5]
```diff
@@ -115,6 +115,9 @@ bool DebugReport::DebugLogMsg(VkFlags msg_flags, const LogObjectList &objects, c
     VkDebugUtilsMessageTypeFlagsEXT types;
     VkDebugUtilsMessageSeverityFlagsEXT severity;
     DebugReportFlagsToAnnotFlags(msg_flags, &severity, &types);
+    if (!(active_severities & severity) || !(active_types & types)) {
+        return false;  // quick check again to make sure use wants these printed
+    }
```
Comment (arno-lunarg) [inline layers/error_message/logging.cpp:120]:
```suggestion
    if (!(active_msg_severities & msg_severity) || !(active_msg_types & msg_types)) {
        return false;  // quick check again to make sure user wants these printed
    }
```

File: layers/error_message/logging.h [hunk 6]
```diff
@@ -219,8 +219,11 @@ class DebugReport {
         return FormatHandle(VkHandleInfo<T>::Typename(), HandleToUint64(handle));
     }
 
+    // Formats messages to be in the proper format, handles VUID logic, and any legacy issues
     bool LogMsg(VkFlags msg_flags, const LogObjectList &objects, const Location &loc, std::string_view vuid_text,
                 const char *format, va_list argptr);
+    // Core logging that interacts with the DebugCallbacks
+    bool DebugLogMsg(VkFlags msg_flags, const LogObjectList &objects, const char *message, const char *text_vuid) const;
```
Comment (arno-lunarg) [inline layers/error_message/logging.h:226]: For consistency with `msg_flags`
```suggestion
    bool DebugLogMsg(VkFlags msg_flags, const LogObjectList &objects, const char *msg, const char *text_vuid) const;
```

**Review by arno-lunarg — APPROVED** [review]

**Pull request merged**